### PR TITLE
attempt to real resolve when there is a quorum failure on reads

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
attempt to real resolve when there is a quorum failure on reads

## Motivation and Context
adds CI/CD test corpus to test with a sample data

## How to test this PR?
CI/CD would cover the case. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [x] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/erasure-object.go [hunk 1]
```diff
@@ -424,7 +440,27 @@ func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object s
...
-	readQuorum, _, err := objectQuorumFromMeta(ctx, metaArr, errs, er.defaultParityCount)
+	readQuorum, _, err := func() (int, int, error) {
+		readQuorum, writeQuorum, versionIDHeal, err := objectQuorumFromMeta(ctx, metaArr, errs, er.defaultParityCount)
+		if err != nil && errors.Is(err, errErasureReadQuorum) {
+			if versionIDHeal != "" && opts.VersionID == "" {
+				// VersionIDHeal is present means it is inconsistent
+				// and when an exact versionID is not specified, we perform heal
+				// based resolution.
+				_, herr := er.HealObject(ctx, bucket, object, versionIDHeal, madmin.HealOpts{
+					NoLock: true,
```
Comment (klauspost) [inline cmd/erasure-object.go]: You cannot heal without locking. You likely only have a read lock here and risk corrupting objects.

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -82,11 +82,27 @@ func (er erasureObjects) CopyObject(ctx context.Context, srcBucket, srcObject, d
...
-	}
+	readQuorum, writeQuorum, err := func() (int, int, error) {
+		readQuorum, writeQuorum, versionIDHeal, err := objectQuorumFromMeta(ctx, metaArr, errs, er.defaultParityCount)
+		if err != nil && errors.Is(err, errErasureReadQuorum) {
+			if versionIDHeal != "" && srcOpts.VersionID == "" {
+				// VersionIDHeal is present means it is inconsistent
+				// and when an exact versionID is not specified, we perform heal
+				// based resolution.
+				_, herr := er.HealObject(ctx, srcBucket, srcObject, versionIDHeal, madmin.HealOpts{
+					NoLock: true,
```
Comment (klauspost) [inline cmd/erasure-object.go]: Same here. Source is only read locked.

File: cmd/erasure-metadata.go [hunk 3]
```diff
@@ -300,6 +300,9 @@ func findFileInfoInQuorum(ctx context.Context, metaArr []FileInfo, modTime time.
 			// make sure that length of Data is same
 			fmt.Fprintf(h, "%v", len(meta.Data))
 
+			// Make sure we get FileInfo with equal number of versions.
+			fmt.Fprintf(h, "%d", meta.NumVersions)
+
```
Comment (klauspost) [inline cmd/erasure-metadata.go]: This seems problematic. You will never heal an older version in your checks, so older inconsistent versions will make it impossible to reach quorum for the current version, even if it is ok.

This can AFAICT give significant regressions.

**Review by klauspost — DISMISSED** [review]

File: cmd/erasure-metadata.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/erasure-metadata.go]: This looks like a flyby change I would have to remove

File: cmd/erasure-object.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/erasure-object.go]: The reason locking would work is that WRITEs are blocked during READs of course you can have as many reads as necessary - since xl.meta is self-contained here.

So no one is really overriding the data here and Heal only heals relevant parts anyways.

File: cmd/erasure-object.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/erasure-object.go]: same as above

**Comment by harshavardhana** [comment]
after talking to @klauspost - will re-implement this differently using the ListObjects technique instead.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#441a1a33-7d9e-4167-999d-879a4fcf4559) |
| mint-gateway-s3.sh | :x: [more...](#3267713e-7a27-43ea-888d-3bdc6d00f1cd) |
| mint-erasure.sh | :x: [more...](#802e86c4-cace-42a5-83ef-448a9ce26d31) |
| mint-pools.sh | :x: [more...](#84b759ae-b738-4bd8-8e5d-d392ec6ba011) |
<a name="802e86c4-cace-42a5-83ef-448a9ce26d31"></a>
### 14613-7e3b985/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:30000
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 22 seconds
{
  "name": "awscli",
  "duration": 1873,
  "function": "aws --endpoint-url http://minio-k8s4:30000 s3api copy-object --bucket awscli-mint-test-bucket-3886 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-3886/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected f90b8f9e401c9d9f4732cc6386646a23, got 69b1f3ab7b77abede10d3b66baa130d8"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 27 seconds
(7/15) Running minio-dotnet tests ... done in 47 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 6 seconds
(9/15) Running minio-java tests ... done in 19 seconds
(10/15) Running minio-js tests ... FAILED in 50 seconds
{
  "name": "minio-js",
  "function": "setBucketVersioning(bucketName, versionConfig):",
  "args": "bucketName:minio-js-test-version-500b5b5a-d215-42f1-9b1f-4ba3ce15301b,versionConfig:{Status:\"Suspended\"}",
  "duration": 6,
  "status": "FAIL",
  "error": "S3Error: We encountered an internal error, please try again.: cause(msgp: attempted to decode type &#34;int&#34; with method for &#34;str&#34;) at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:165:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:89:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:670:10) at endReadableNT (internal/streams/readable.js:1334:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/15) Running minio-py tests ... done in 1 minutes and 0 seconds
(11/15) Running s3cmd tests ... done in 25 seconds
(12/15) Running s3select tests ... FAILED in 0 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-e3925fa7-4e40-413d-9cf2-3835e0917d55"
  },
  "duration": 127,
  "message": "Test test_4: data mismatch. Expected : b'{\"_1\":\"\\\\\"col1\\\\\"\",\"_2\":\"col2\",\"_3\":\"col3\"}\\n', Received b'{\"_1\":\"l2\",\"_2\":\"col3\"}\\n{\"_1\":\"\\\\u0000l2\",\"_2\":\"col3\"}\\n'",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 117, in test_csv_input_custom_quote_char\n    test_sql_api(f'test_{idx}', client, bucket_name,\n  File \"/mint/run/core/s3select/csv.py\", line 54, in test_sql_api\n    raise ValueError('Test {}: data mismatch. Expected : {}, Received {}'.format(\nValueError: Test test_4: data mismatch. Expected : b'{\"_1\":\"\\\\\"col1\\\\\"\",\"_2\":\"col2\",\"_3\":\"col3\"}\\n', Received b'{\"_1\":\"l2\",\"_2\":\"col3\"}\\n{\"_1\":\"\\\\u0000l2\",\"_2\":\"col3\"}\\n'\n",
  "status": "FAIL"
}
(12/15) Running versioning tests ... done in 3 minutes and 2 seconds

Executed 12 out of 15 tests successfully.
```
<a name="441a1a33-7d9e-4167-999d-879a4fcf4559"></a>
### 14613-7e3b985/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s7:31122
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 3 seconds
(3/15) Running aws-sdk-php tests ... done in 46 seconds
(4/15) Running aws-sdk-ruby tests ... FAILED in 14 seconds
{
  "name": "aws-sdk-ruby",
  "function": "presignedGet(bucket_name,file_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-5c93e726e607",
    "file_name": "datafile-1-kB"
  },
  "duration": 159.92,
  "error": "Check sum values do NOT match",
  "status": "FAIL"
}
(4/15) Running awscli tests ... done in 1 minutes and 5 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 33 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 12 seconds
(8/15) Running minio-go tests ... done in 2 minutes and 23 seconds
(9/15) Running minio-java tests ... FAILED in 9 seconds
{
  "name": "minio-java",
  "function": "getBucketVersioning()",
  "duration": 96,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InternalError, message = We encountered an internal error, please try again.: cause(loadBucketMetadata: unknown format: 6700), bucketName = minio-java-test-59gdi9, objectName = null, resource = /minio-java-test-59gdi9, requestId = 16E4225CA9AE4552, hostId = 1773a88a-ea37-4217-a216-878297b7fed8)\nrequest={method=PUT, url=http://minio-k8s7:31122/minio-java-test-59gdi9?versioning=, headers=Host: minio-k8s7:31122\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.3.6\nContent-MD5: QQFYoy/mRYV9PGZUfFi0Bw==\nx-amz-content-sha256: 482d4e510b5cdcf0bbf3a044e82832dc711f4f28757a6c06eb21758282a73ef7\nx-amz-date: 20220409T051224Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220409/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Length: 382\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 16E4225CA9AE4552\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Sat, 09 Apr 2022 05:12:24 GMT\n}\n >>> [io.minio.S3Base.execute(S3Base.java:645), io.minio.S3Base.execute(S3Base.java:440), io.minio.S3Base.executePut(S3Base.java:772), io.minio.MinioClient.setBucketVersioning(MinioClient.java:1236), FunctionalTest.getBucketVersioning(FunctionalTest.java:650), FunctionalTest.runBucketTests(FunctionalTest.java:3710), FunctionalTest.runTests(FunctionalTest.java:3780), FunctionalTest.runEndpointTests(FunctionalTest.java:3880), FunctionalTest.main(FunctionalTest.java:4001)]"
}
(9/15) Running minio-js tests ... done in 1 minutes and 35 seconds
(10/15) Running minio-py tests ... done in 1 minutes and 56 seconds
(11/15) Running s3cmd tests ... done in 28 seconds
(12/15) Running s3select tests ... done in 10 seconds
(13/15) Running versioning tests ... done in 3 minutes and 15 seconds

Executed 13 out of 15 tests successfully.
```
<a name="84b759ae-b738-4bd8-8e5d-d392ec6ba011"></a>
### 14613-7e3b985/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:31859
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 45 seconds
(4/15) Running aws-sdk-ruby tests ... done in 11 seconds
(5/15) Running awscli tests ... done in 2 minutes and 8 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 34 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 21 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 23 seconds
(10/15) Running minio-java tests ... FAILED in 1 minutes and 16 seconds
{
  "name": "minio-java",
  "function": "removeBucket()",
  "duration": 20,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-16mof99, objectName = null, resource = /minio-java-test-16mof99, requestId = 16E422881820E4F5, hostId = a01272e4-c128-442c-81bf-4f8b7421cfd6)\nrequest={method=DELETE, url=http://minio-k8s4:31859/minio-java-test-16mof99, headers=Host: minio-k8s4:31859\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.3.6\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20220409T051531Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220409/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 328\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 16E422881820E4F5\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Sat, 09 Apr 2022 05:15:31 GMT\n}\n >>> [io.minio.S3Base.execute(S3Base.java:645), io.minio.S3Base.execute(S3Base.java:440), io.minio.S3Base.executeDelete(S3Base.java:745), io.minio.MinioClient.removeBucket(MinioClient.java:1592), FunctionalTest.teardown(FunctionalTest.java:714), FunctionalTest.runObjectTests(FunctionalTest.java:3776), FunctionalTest.runTests(FunctionalTest.java:3781), FunctionalTest.runEndpointTests(FunctionalTest.java:3880), FunctionalTest.main(FunctionalTest.java:4001)]"
}
(10/15) Running minio-js tests ... done in 1 minutes and 29 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 14 seconds
(12/15) Running s3cmd tests ... done in 30 seconds
(13/15) Running s3select tests ... FAILED in 2 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-94c24768-6fdf-4c96-bbef-5f3ebafdc325"
  },
  "duration": 2095,
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-94c24768-6fdf-4c96-bbef-5f3ebafdc325, request_id: 16E422C38DE6CD39, host_id: a01272e4-c128-442c-81bf-4f8b7421cfd6, bucket_name: s3select-test-94c24768-6fdf-4c96-bbef-5f3ebafdc325",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 120, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 669, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 397, in _execute\n    return self._url_open(\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 380, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-94c24768-6fdf-4c96-bbef-5f3ebafdc325, request_id: 16E422C38DE6CD39, host_id: a01272e4-c128-442c-81bf-4f8b7421cfd6, bucket_name: s3select-test-94c24768-6fdf-4c96-bbef-5f3ebafdc325\n",
  "status": "FAIL"
}
(13/15) Running versioning tests ... done in 3 minutes and 27 seconds

Executed 13 out of 15 tests successfully.
```
<a name="3267713e-7a27-43ea-888d-3bdc6d00f1cd"></a>
### 14613-7e3b985/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s5:31948
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... FAILED in 23 seconds
{
  "name": "awscli",
  "duration": 8608,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-17594/datafile-1-kBdelete: s3://awscli-mint-test-bucket-17594/awscli-mint-test-bucket-17594-objectremove_bucket failed: s3://awscli-mint-test-bucket-17594 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 29 seconds
(7/15) Running minio-dotnet tests ... done in 28 seconds
(8/15) Running minio-go tests ... done in 41 seconds
(9/15) Running minio-java tests ... done in 42 seconds
(10/15) Running minio-js tests ... done in 47 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 7 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 7 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running versioning tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
The functionality now is fully generalized and supports all the corner cases @klauspost @kannappanr PTAL

**Comment by harshavardhana** [comment]
PTAL @klauspost @kannappanr @krisis @poornas

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL @klauspost since you are back your feedback is needed in this PR.

**Comment by klauspost** [comment]
@harshavardhana Oh, wow.. Yeah - will look at it. Review ready when you get up!

File: cmd/storage-datatypes.go [hunk 4]
```diff
@@ -112,9 +112,24 @@ func (f *FileInfoVersions) findVersionIndex(v string) int {
 	return -1
 }
 
+// RawFileInfo - represents raw file stat information as byte array.
+//msgp:tupe RawFileInfo
```
Comment (klauspost) [inline cmd/storage-datatypes.go]:
```suggestion
```
(misspelled anyways, and you have short keys, so overhead is max a few bytes)

**Review by klauspost — APPROVED** [review]
Beautiful!

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally
